### ISSUE TITLE
contains_eager() .propagate_to_loaders doesn't work for a composite load option

### ISSUE THREAD

**Issue by awgymer** [sqlalchemy/sqlalchemy#6449]
**Describe your question**

I have some code where I query for all instances of a model with some grandchildren `joinedload`. When I run an `update` on that model it then throws an error.

```py
DBBase = declarative_base()

class UserInventory(DBBase):
    user_inventory_id = Column(Integer, primary_key=True)
    inventory_id = Column(
        Integer, ForeignKey('inventory.inventory_id'),
        nullable=False
    )
    is_checked = Column(Boolean, default=False)
    inventory = relationship('Inventory', back_populates='user_sets')

class UserPart(DBBase):
    user_part_id = Column(Integer, primary_key=True)
    user_inventory_id = Column(
        Integer,
        ForeignKey('user_inventory.user_inventory_id'),
        nullable=False
    )
    inventory_part_id = Column(
        Integer,
        ForeignKey('inventory_part.id'),
        nullable=False
    )
    num_missing = Column(Integer, default=0, server_default='0')
    part = relationship('InventoryPart')
    user_set = relationship('UserInventory', back_populates='missing')
    __table_args__ = (
        UniqueConstraint(user_inventory_id, inventory_part_id, name='mp_uniq'),
    )

class InventoryPart(DBBase):
    id = Column(Integer, primary_key=True)
    inventory_id = Column(
        Integer, ForeignKey('inventory.inventory_id', ondelete='CASCADE'),
        nullable=False
    )
    part_id = Column(
        TEXT, ForeignKey('part.part_id'),
        nullable=False
    )
    color_id = Column(
        Integer, ForeignKey('color.color_id')
    )
    part = relationship('Part', overlaps='element, part_id')
    color = relationship('Color', overlaps='element, color_id')

UI_ID = 1

parts = session.query(UserPart).\
        join(InventoryPart).\
        filter(UserPart.user_inventory_id == UI_ID).\
        order_by(
            InventoryPart.is_spare,
            InventoryPart.color_id
        ).\
        options(
            contains_eager(UserPart.part).joinedload(InventoryPart.part),
            contains_eager(UserPart.part).joinedload(InventoryPart.color)
        ).all()

session.query(UserPart).filter(UserPart.user_inventory_id == UI_ID).\
                update({"num_missing": 0}, synchronize_session="fetch")
session.commit()
```

**Complete stack trace, if applicable**
It throws this error on `session.commit()`
```
/Users/<me>/.pyenv/versions/env1/lib/python3.8/site-packages/sqlalchemy/orm/strategies.py:2580: SAWarning: Multiple rows returned with uselist=False for eagerly-loaded attribute 'UserPart.part' 
  util.warn(
```

I tried removing the `synchronize_session` option and it changed nothing. If I run the update and *then* run my `parts`
query, then no issues. It obviously has something to do with the joins before the update I think? 

**Versions**
 - OS: Mac Catalina 10.15.2
 - Python: 3.8.2
 - SQLAlchemy: 1.4.11
 - Database: SQLite
 - DBAPI:

**Comment by zzzeek** [comment]
hi there -

I have no idea what is happening here, there's no stack trace or illustration of the data being inserted, updated, or selected.     the warning means quite simply that there are multiple rows being returned on behalf of UserPart.part which would indicate that query(UserPart).join(InventoryPart) is not joining in the way you'd expect, but I have absolutely no idea how that would occur in the "commit", the commit would not run any query here.

I would need full SQL log output, complete stack trace, test data to show the problem.   Browse the issues here for the [great mcve](https://github.com/sqlalchemy/sqlalchemy/issues?q=is%3Aissue+label%3A%22great+mcve%22+is%3Aclosed) for examples on how to provide a proper [MCVE](http://stackoverflow.com/help/mcve) so we can see what exactly is happening here.

**Comment by zzzeek** [comment]
I will at least note that I doubt SQLAlchemy knows how to eager load this correctly:

```
           contains_eager(UserPart.part).joinedload(InventoryPart.part),
            contains_eager(UserPart.part).joinedload(InventoryPart.color)
```

I've never seen that attempted before, I have no idea if that works or not.

**Comment by awgymer** [comment]
Ok so you're right, it can't be replicated with my code. I have just discovered that it doesn't actually error on the `commit` but rather when you then try to access the results in the query *after* the update. 

The below should be a fully reproducible example. 

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy.orm import contains_eager
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session

Base = declarative_base()


class Part(Base):
    __tablename__ = "part"
    id = Column(Integer, primary_key=True)


class Color(Base):
    __tablename__ = "color"
    id = Column(Integer, primary_key=True)


class IPart(Base):
    __tablename__ = "ipart"
    id = Column(Integer, primary_key=True)
    p_id = Column(ForeignKey("part.id"))
    c_id = Column(ForeignKey("color.id"))
    color = relationship('Color')
    part = relationship('Part')


class UPart(Base):
    __tablename__ = "upart"
    id = Column(Integer, primary_key=True)
    i_id = Column(ForeignKey("ipart.id"))
    num_missing = Column(Integer, server_default='0')
    part = relationship('IPart')


e = create_engine("sqlite:///./test.sqlite", echo=True)


Base.metadata.create_all(e)

s = Session(e)

p1 = Part()
c1 = Color()
c2 = Color()

s.add(p1)
s.add(c1)
s.add(c2)
s.commit()

i1 = IPart(p_id=p1.id, c_id=c2.id)
i2 = IPart(p_id=p1.id, c_id=c1.id)

s.add(i1)
s.add(i2)
s.commit()

u1 = UPart(i_id=i1.id)
u2 = UPart(i_id=i1.id)
u3 = UPart(i_id=i2.id)
s.add(u1)
s.add(u2)
s.add(u3)
s.commit()
s.close()

s = Session(e)

q = s.query(UPart).join(IPart).\
    filter(UPart.id != 1).\
    order_by(IPart.c_id).\
    options(
        contains_eager(UPart.part).joinedload(IPart.color),
        contains_eager(UPart.part).joinedload(IPart.part)
    )

r = q.all()

print(r[0].num_missing)
print(r[0].part.color.id)

update_query = s.query(UPart).\
    filter(UPart.id != 1).update(
    {"num_missing": 1}, synchronize_session="fetch"
)

s.commit()

print(r[0].part)
```

Note that closing the session in which you create the objects and opening a new session to run the queries in seems to be key in reproducing this bug. In my real-world example obviously the models I am retrieving from the DB were put there by totally separate code. 

I'll also include the last few lines of output.

```
2021-05-08 19:11:21,881 INFO sqlalchemy.engine.Engine SELECT upart.id 
FROM upart 
WHERE upart.id != ?
2021-05-08 19:11:21,881 INFO sqlalchemy.engine.Engine [generated in 0.00019s] (1,)
2021-05-08 19:11:21,882 INFO sqlalchemy.engine.Engine UPDATE upart SET num_missing=? WHERE upart.id != ?
2021-05-08 19:11:21,882 INFO sqlalchemy.engine.Engine [generated in 0.00009s] (1, 1)
2021-05-08 19:11:21,882 INFO sqlalchemy.engine.Engine COMMIT
2021-05-08 19:11:21,883 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-05-08 19:11:21,884 INFO sqlalchemy.engine.Engine SELECT ipart.id AS ipart_id, ipart.p_id AS ipart_p_id, ipart.c_id AS ipart_c_id, upart.id AS upart_id, upart.i_id AS upart_i_id, upart.num_missing AS upart_num_missing, color_1.id AS color_1_id, part_1.id AS part_1_id 
FROM ipart, upart LEFT OUTER JOIN color AS color_1 ON color_1.id = ipart.c_id LEFT OUTER JOIN part AS part_1 ON part_1.id = ipart.p_id 
WHERE upart.id = ?
2021-05-08 19:11:21,884 INFO sqlalchemy.engine.Engine [generated in 0.00009s] (3,)
/Users/arthurgymer/.pyenv/versions/lego/lib/python3.8/site-packages/sqlalchemy/orm/strategies.py:2580: SAWarning: Multiple rows returned with uselist=False for eagerly-loaded attribute 'UPart.part' 
  util.warn(
<__main__.IPart object at 0x104c328b0>
```


As for the:
```
contains_eager(UserPart.part).joinedload(InventoryPart.part),
contains_eager(UserPart.part).joinedload(InventoryPart.color)
```
 It seems to work fine. It's not the issue here either so far as I can tell, I removed it and nothing changed. My aim with that code was to eager-load *both* the `part` and `color` relationships on the eagerly-loaded `UserPart.part`. The docs show use of chained `joinedload` for 1 eagerly loaded child and 1 eagerly loaded grandchild, but not for 2 grandchildren. I tried a few things but this was the only one that seemed to work. The SQL issued by the query looks reasonable too.

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **master** branch:

**Ensure propagate_to_loaders honored at the sub-loader level** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2813

**Comment by zzzeek** [comment]
thanks, this is a regression caused by #1763 and will the fix will be in 1.4.15.
